### DISCUSSION TITLE
A Commit strategy failed exception occurs when a transaction is submitted and txId is obtained

### DISCUSSION THREAD

**Discussion by liow10** [hyperledger/fabric-gateway-java#169] [Q&A]
The Commit strategy failed exception is thrown when a transaction is submitted using the fabric gateway. However, the data submitted this time can be checked when query is invoked, but the transaction hash cannot be obtained. If I use transaction.evaluate(), Then the submitted data will not be recorded on the fabric. How can I solve this problem?
![Snipaste_2024-02-23_15-48-01](https://github.com/hyperledger/fabric-gateway-java/assets/157773725/1cf6a4af-8504-4c44-821b-04fb292fef9b)
![企业微信截图_17086745635177](https://github.com/hyperledger/fabric-gateway-java/assets/157773725/da403a11-408c-4d41-847d-218730fef4d5)
![image](https://github.com/hyperledger/fabric-gateway-java/assets/157773725/bb3fb617-46fa-4a11-a012-9edd4c2ce82e)

**Comment by bestbeforetoday** [comment]
Please note that it is much more helpful to [quote](https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#quoting-code) code / stacktrace snippets directly in the text rather than embed images of them. The text is much more readable and also searchable, whereas images are neither.

From the disconnect being handled in the stacktrace, it looks like your client's block eventing connection to a peer failed. After this disconnect is handled, the commit strategy is not waiting for responses from any more peers. No valid transactions responses have been received from any other peers so the strategy fails. The commit strategy failure is similar to a timeout at the same stage: there is no clear indication of either transaction failure or transaction success so the state of the transaction in the Fabric network is unknown.

The cause seems to be a (block) eventing connection failure between your client and peer(s). You might find some logs either at the client or peer end that give more information on why the disconnection occurred. One possibility is that something in the network path is closing idle connections. If this is the case, you might need to apply appropriate gRPC keep-alive settings to ensure the connection is not idle long enough for that to happen.

As an aside, you can get the transaction ID from the client-side Transaction object (using [getTransactionId](https://hyperledger.github.io/fabric-gateway-java/main/org/hyperledger/fabric/gateway/Transaction.html#getTransactionId())) even before submitting the transaction so there is no need to return it from the smart contract transaction function.

Finally, please consider moving to Fabric v2.5 and the newer [Fabric Gateway client API](https://hyperledger.github.io/fabric-gateway/), which supersedes the fabric-gateway-java API.
